clients/src/main/java/org/apache/kafka/clients/producer/Callback.java

```diff
@@ -25,10 +25,12 @@
     /**
      * A callback method the user can implement to provide asynchronous handling of request completion. This method will
      * be called when the record sent to the server has been acknowledged. When exception is not null in the callback,
-     * metadata will contain the special -1 value for all fields except for topicPartition, which will be valid.
+     * and the exception is a subclass of ApiException, metadata will be null. For all other exceptions, metadata will
```

Comment:
It seems that the fix in https://github.com/apache/kafka/pull/5798 was incorrect. The changes for the comment were intended for InterceptorCallback.onComplete. For Callback.onCompletion, currently, if there is an exception, it seems that metadata will always be null.

For now, we could probably just revert the comment on Callback.onCompletion. 

It's probably useful to make the behavior consistent between Callback.onCompletion and InterceptorCallback.onComplete. We could file a jira to track that since it probably needs a KIP.

Reply:
I don't think metadata will always be null, as the updated metadata will sometimes be passed to the user callback as far as I can see from the InterceptorCallback implementation (see [here](https://github.com/apache/kafka/blob/3.1/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1387-L1390)).

I believe we have been seeing a mixture of null metadata + placeholder metadata in our Kafka deployment (v2.3.1), which has caused some issues for my team.

Reply:
@soceanainn : Thanks for the comment. You are right. It seems that we are generating metadata inconsistently. It seems that if the send() call throws an ApiException, we return a null metadata. Otherwise, the callback will always get a placeholder metadata when exception is not null, even if the exception is of ApiException.

Reply:
What do you suggest? Should the docs be updated, or perhaps we should simply remove lines [1001-1002](https://github.com/apache/kafka/blob/3.1/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1001-L1002), as this would make the behaviour consistent across `Callback.onCompletion` and `InterceptorCallback.onComplete`, and in line with the current documented behaviour